s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+
+    virtual future<> close() override {
+        return make_ready_future<>();
+    }
+
+    virtual std::unique_ptr<seastar::file_handle_impl> dup() override {
+        unsupported(__FUNCTION__); // FIXME
+    }
+
+    virtual future<temporary_buffer<uint8_t>> dma_read_bulk(uint64_t offset, size_t size, const io_priority_class& pc) override {
+        auto buf = co_await _client->get_object(_path, offset, size);
+        co_return temporary_buffer<uint8_t>(reinterpret_cast<uint8_t*>(buf.get_write()), buf.size(), buf.release());
+    }
+
+    virtual future<size_t> read_dma(uint64_t pos, void* buffer, size_t len, const io_priority_class& pc) override {
+        unsupported(__FUNCTION__); // FIXME
+    }
+
+    virtual future<size_t> read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc) override {
+        unsupported(__FUNCTION__); // FIXME
+    }
+};
+
+connection_ptr::~connection_ptr() {
+    if (!_ptr || !_ptr.owned() || _ptr->closed()) {
+        return;
+    }
+    auto ptr = std::move(_ptr);
+    log.debug("conn {}: not closed, closing in background", fmt::ptr(&*ptr));
+    (void)ptr->close()
+        .then([ptr] {
+            log.debug("conn {}: closed", fmt::ptr(&*ptr));
+        }).handle_exception([ptr](auto ep) {
+            log.error("conn {}: failed to close: {}", fmt::ptr(&*ptr), ep);
+        });
+}
+
+class client_impl : public client, public enable_shared_from_this<client_impl> {
+    connection_factory_ptr _connection_factory;
+    std::map<sstring, sstring> _headers;
+protected:
+    future<connection_ptr> new_connection() {
+        return _connection_factory->connect();
+    }
+
+    template <typename Func, typename FuncResult = std::invoke_result_t<Func, connection_ptr>>
+    futurize_t<FuncResult>
+    do_with_connection(Func func) {
+        return new_connection().then([func = std::move(func), this] (connection_ptr con) mutable {
+            return futurize_invoke(std::move(func), con)
+                    .then_wrapped([con, this] (auto&& f) mutable {
+                        if (!f.failed()) {
+                            _connection_factory->take_back(con);
+                            return std::move(f);
+                        } else {
+                            return con->close()
+                                .handle_exception([con] (auto ep) mutable {
+                                    log.error("conn {}: failed to close: {}", fmt::ptr(&*con), ep);
+                                }).then([con, f = std::move(f)] () mutable {
+                                    return std::move(f);
+                                });
+                        }
+                    });
+        });
+    }
+private:
+    class s3_upload_data_sink : public data_sink_impl {
+        // "Each part must be at least 5 MB in size, except the last part."
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        static constexpr size_t minimum_part_size = 5*1024*1024;
```

Comment:
Is there a reason not to upload it in one huge chunk?

Reply:
"Upload an object in a single operation using the AWS SDKs, REST API, or AWS CLI—With a single PUT operation, you can upload a single object up to 5 GB in size.". Better to document it.